### ISSUE TITLE
createMock fails with internal class \SimpleXMLElement

### ISSUE THREAD

**Issue by bug-tape** [sebastianbergmann/phpunit-mock-objects#417]
| Q                            | A
| -----------------------------| ---------------
| phpunit-mock-objects version | 6.1.1
| PHPUnit version              | 7.1.3
| PHP version                  | 7.2.3
| Installation Method          | Composer

After upgrading phpunit-mock-objects from version 6.0.1 using createMock with \SimpleXMLElement stopped working with the following message:
Mock_SimpleXMLElement_4670bd5b::__phpunit_setReturnValueGeneration(): Node no longer exists

**Comment by sebastianbergmann** [comment]
#### Reproducing Example
```php
<?php declare(strict_types=1);
use PHPUnit\Framework\TestCase;

final class Test extends TestCase
{
    public function testOne(): void
    {
        $this->createMock(SimpleXMLElement::class);
    }
}
```
```
$ phpunit Test
PHPUnit 7.1.3 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 27 ms, Memory: 4.00MB

There was 1 error:

1) Test::testOne
Mock_SimpleXMLElement_75717022::__phpunit_setReturnValueGeneration(): Node no longer exists

/home/sb/Test.php:8

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

**Comment by sebastianbergmann** [comment]
CC @lstrojny

**Comment by sebastianbergmann** [comment]
Below is a stripped-down version of the code generated by `PHPUnit\Framework\MockObject\Generator` for this:

```php
<?php
class Mock_SimpleXMLElement_23b3064f extends SimpleXMLElement
{
    private $__phpunit_returnValueGeneration = true;

    public function __phpunit_setReturnValueGeneration(bool $returnValueGeneration)
    {
        $this->__phpunit_returnValueGeneration = $returnValueGeneration;
    }
}

$rc = new ReflectionClass(Mock_SimpleXMLElement_23b3064f::class);
$o  = $rc->newInstanceWithoutConstructor();

$o->__phpunit_setReturnValueGeneration(false);
```
```
$ php test.php
PHP Warning:  Mock_SimpleXMLElement_23b3064f::__phpunit_setReturnValueGeneration(): Node no longer exists in /home/sb/test.php on line 8
```
Apparently `SimpleXMLElement` is "special" and extending it is problematic. It's weird, though, that this (supposedly) worked before `__phpunit_setReturnValueGeneration()` was introduced.

**Comment by navarr** [comment]
It looks like the problem is that SimpleXMLElement is intercepting all calls to properties and running them through attempting to get child nodes?

Doing something like:
```
$rc = new ReflectionClass(SimpleXMLElement::class);
$o = $rc->newInstanceWithoutConstructor();
$o->testNode = true;
```

Returns the same error, but of course
```
$e = new SimpleXMLElement('<test />');
$e->testNode = true;
```

does not.

I'm guessing some part of the construction of the element is important - probably creating some sort of DOMNode for properties to be assigned against, and then catching all of them and assigning them to that DOM node - is my best guess.

Without the node being created, the properties can't be assigned - and we get this error?

**Comment by SenseException** [comment]
Creating a mock of SPLFileInfo with disabled constructor once didn't worked either. Maybe we have a similar case?

5.3 worked, 5.4 and 5.5 not:

https://speakerdeck.com/senseexception/spl-lost-in-documentation#82

**Comment by bug-tape** [comment]
@sebastianbergmann You are right: createMock already did not work with older versions because of same problem. Our tests do not create a mock directly, but we have some methods which have \SimpleXMLElement as return type hint. And using a mock for classes with such method is failing with the new version.
As workaround we now could manually define a return value with a dummy \SimpleXMLElement instance for these methods in such mocks. This is of cause not satisfying.

**Comment by sebastianbergmann** [comment]
The the feature implemented in #405 might solve your problem. It should allow you to stub methods that have `SimpleXMLElement` in a return type declaration.
